# Exercise 6: NumPy

<font color = hotpink>Jessica Reyes<font>

In [2]:
import numpy as np

### NaN and infinity

NumPy (and other computer languages/libraries) have a special value called `NaN` ([Not a Number](https://en.wikipedia.org/wiki/NaN)).
NaN is the result of various arithmetic operations, e.g. 0/0, $-\infty + \infty$.

In NumPy, NaN is represented as `np.nan`.  NaN satisfies the following properties:

* Comparisons of NaN with a number x using one of the operators =, <, >, >=, and <= is always False.  Consequently, x != NaN is always True.

* Any arithmetic operation involving a NaN leads to a NaN 

Infinity is represented using `np.inf`, and negative infinity is `-np.inf`.

* Experiment with NaN and infinity and convince yourself that the results of the following operations make sense to you:

```Pytnon
0 * np.nan
np.nan == np.nan
np.inf > np.nan
np.nan - 1
```

In [3]:
print(0*np.nan)
print(np.nan == np.nan)
print(np.inf > np.nan)
print(np.nan - 1)

nan
False
False
nan


Now try the following:

In [4]:
print(0.3 == 3 * 0.1)
print(3*0.1)

False
0.30000000000000004


Can you explain the result?

Python represents the floating number (0.1 in binary) is like a repeating fraction. When you multiply 3*0.1 = 0.30000000000000004 as a result rather than 0.3. So it is slightly larger than 0.3. Which is why it evaluates to False.

### Creating arrays

* Create a 3x3 matrix with integer values ranging from 0 to 8.  Do this by creating a one dimensional array and using the [reshape](https://numpy.org/doc/stable/reference/generated/numpy.reshape.html) function.

In [5]:
arr_1 = np.arange(9)
arr_2 = arr_1.reshape(3,3)
print(arr_2)

[[0 1 2]
 [3 4 5]
 [6 7 8]]


* Create the same 3x3 matrix and have the end result be a floating point array.

In [6]:
arr_1 = np.arange(9, dtype = float)
arr_2 = arr_1.reshape(3,3)
print(arr_2)

[[0. 1. 2.]
 [3. 4. 5.]
 [6. 7. 8.]]


* Create a vector of length 10 with random values between 0 and 1

In [7]:
random_vector = np.random.rand(10)
print(random_vector)

[0.36895032 0.45648289 0.53393482 0.52155089 0.93745497 0.47175906
 0.42743008 0.23045741 0.98098383 0.05978002]


* Create a 10x10 array with 1 on the border and 0 inside

In [8]:
arr_3 = np.ones((10,10), dtype=int)
arr_3[1:-1,1:-1] = 0
print(arr_3)

[[1 1 1 1 1 1 1 1 1 1]
 [1 0 0 0 0 0 0 0 0 1]
 [1 0 0 0 0 0 0 0 0 1]
 [1 0 0 0 0 0 0 0 0 1]
 [1 0 0 0 0 0 0 0 0 1]
 [1 0 0 0 0 0 0 0 0 1]
 [1 0 0 0 0 0 0 0 0 1]
 [1 0 0 0 0 0 0 0 0 1]
 [1 0 0 0 0 0 0 0 0 1]
 [1 1 1 1 1 1 1 1 1 1]]


* For an integer $n$, create an array of length $3n$ filled with the cyclic pattern 1, 2, 3.  For example, for $n=2$, the output should be ```[1,2,3,1,2,3]```.

In [9]:
cyclic_array = np.array([1,2,3])
n = 2
print(np.tile(cyclic_array, n))

[1 2 3 1 2 3]


### Slices

Describe the effect of each of the following slices in words.

The first set of slices refer to the following array:
```Python
a = np.arange(10)
```

For example, the slice ```a[0:5]``` yields the first five elements of the array.


Slice 1:
```Python
a[::2]
```

In [10]:
a = np.arange(10)

print(a[::2])

[0 2 4 6 8]


The slice a[0::2] prints out the even elements of the array.

Slice 2:
```Python
a[5:]
```

In [12]:
print(a[5:])

[5 6 7 8 9]


The slice a[5:] prints out the elements starting at 5 to the end of the array.

The next set of questions refer to the following two-dimensional array, i.e. matrix:

```Python
a2d = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
```

Slice 3:
```Python
a2d[0]
```

In [14]:
a2d = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(a2d[0])

[1 2 3]


The slice a2d[0] prints out the first row of the matrix.

Slice 4:
```Python
a2d[1, :2]
```

In [15]:
print(a2d[1, :2])

[4 5]


The slice a2d[1, :2] prints out the first two elements of the 2nd row.

Slice 5:
```Python
a2d[:, 1]
```

In [16]:
print(a2d[:, 1])

[2 5 8]


The slice a2d[:, 1] prints out the second column of the matrix.

Slice 6:
```Python
a2d[:, 1:]
```

In [17]:
print(a2d[:, 1:])

[[2 3]
 [5 6]
 [8 9]]


The slice a2d[:, 1:] prints out the 2nd and last columns of the matrix.

Slice 7:
```Python
a2d[:,:]
```

In [19]:
print(a2d[:,:])

[[1 2 3]
 [4 5 6]
 [7 8 9]]


The slice a2d[:,:] prints out every row and every column (so the entire matrix).

Here's a more challenging exercise that you can solve using slices:

* Create a 8x8 matrix that looks like this:

```
[[0 1 0 1 0 1 0 1]                                                      
 [1 0 1 0 1 0 1 0]                                                      
 [0 1 0 1 0 1 0 1]                                                      
 [1 0 1 0 1 0 1 0]                                                      
 [0 1 0 1 0 1 0 1]                                                      
 [1 0 1 0 1 0 1 0]                                                      
 [0 1 0 1 0 1 0 1]                                                      
 [1 0 1 0 1 0 1 0]]
```

Hint:  even and odd numbered rows each are filled with the same pattern.


In [20]:
eight_by_eight = np.zeros((8,8), dtype=int)
eight_by_eight[1::2,::2] = 1
eight_by_eight[::2,1::2] = 1
print(eight_by_eight)

[[0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]
 [0 1 0 1 0 1 0 1]
 [1 0 1 0 1 0 1 0]]
